core/server/master/src/main/java/alluxio/master/file/meta/InvalidationSyncCache.java

```diff
@@ -0,0 +1,418 @@
+                isFile ? syncState.mDirectValidationTime :
+                descendantType != DescendantType.NONE ? syncState.mRecursiveValidationTime
+                    : syncState.mDirectValidationTime);
+            break;
+          default: // at a higher ancestor path
+            lastValidationTime = Math.max(lastValidationTime, syncState.mRecursiveValidationTime);
+        }
+      }
+      if (currPath.equals(AlluxioURI.SEPARATOR)) {
+        break;
+      }
+      currPath = PathUtils.getParent(currPath);
+      parentLevel++;
+    }
+    long currentTime = mClock.millis();
+    SyncCheck result;
+
+    if (intervalMs == 0) {
+      // Always sync.
+      return SyncCheck.shouldSyncWithTime(lastValidationTime);
+    }
+    if (lastInvalidationTime == 0 && intervalMs < 0) {
+      // if no invalidation, and a negative interval then do not sync
+      result = SyncCheck.shouldNotSyncWithTime(lastValidationTime);
+    } else if (lastInvalidationTime >= lastValidationTime) {
+      // if a recent invalidation then we always sync
+      result = SyncCheck.shouldSyncWithTime(lastValidationTime);
+    } else if (intervalMs < 0) {
+      // do not sync with a negative interval
+      result = SyncCheck.shouldNotSyncWithTime(lastValidationTime);
+    } else if ((currentTime - lastValidationTime) >= intervalMs) {
+      // syncing is needed based on an interval
+      result = SyncCheck.shouldSyncWithTime(lastValidationTime);
+    } else {
+      // syncing is not needed based on an interval
+      result = SyncCheck.shouldNotSyncWithTime(lastValidationTime);
+    }
+    LOG.debug("Result of should sync path {}: {}, invalidation time: {}, validation time: {},"
+            + " clock time: {}",
+        result, path, lastInvalidationTime, lastValidationTime, currentTime);
+    return result;
+  }
+
+  /**
+   * Notify that a path has been invalidated.
+   * @param path the path
+   */
+  @VisibleForTesting
+  public void notifyInvalidation(AlluxioURI path) throws InvalidPathException {
+    String currPath = path.getPath();
+    long time = mClock.millis();
+    notifyInvalidationInternal(currPath, time);
+  }
+
+  private void notifyInvalidationInternal(String currPath, long time)
+      throws InvalidPathException {
+    LOG.debug("Set sync invalidation for path {} at time {}", currPath, time);
+    int parentLevel = 0;
+
+    if (currPath.equals(AlluxioURI.SEPARATOR)) {
+      try (LockResource ignored = new LockResource(mRootLock)) {
+        mRoot.setInvalidationTime(time);
+      }
+    } else {
+      mItems.asMap().compute(currPath, (key, state) -> {
+        if (state == null) {
+          state = new SyncState(false);
+        }
+        // set the invalidation time for the path
+        state.setInvalidationTime(time);
+        return state;
+      });
+    }
+    while (!currPath.equals(AlluxioURI.SEPARATOR)) {
+      currPath = PathUtils.getParent(currPath);
+      parentLevel++;
+      if (currPath.equals(AlluxioURI.SEPARATOR)) {
```

Comment:
i haven't thought too much so please correct me if i'm wrong. If we recursively make all parents invalidated, will invalidation on `/a/b/c/d/e` cause `ls -R /` to have a huge recursive sync? I mean invalidation on `/a/b/c/d/e` should only trigger the sync on itself ideally, right?

If the sync-amplification problem does exist, I guess one alternative is we don't set the invalidation marker on parents, but instead when we are listing, check each child if a smaller-scoped sync should be triggered. Is that possible?

Reply:
Yes, the invalidation to let the parent know that a sync is needed, so a new sync will be launched.

But when checking each child path it will recheck that path individually to see if a sync is needed here https://github.com/tcrain/alluxio/blob/invalidation-cache-timestamp/core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java#L666-L671 and only sync the paths that are needed.

There is still extra work being done by calling list status on each marked path, so I think it can be improved further, but I think this might need a much more complex data structure. For example an actually tree (trie) might work a lot better instead of a path cache. For now I think sticking with the current path cache is good as it fits in the current design, then we can rethink the whole design when we rewrite metadata sync.

Reply:
Not sure I fully caught up so pls correct me if I'm wrong. 

1. When we invalidate a path, we mark the invalidation timestamp all the way up to the root. 
2. And later when we `ls` a parent path, an `InodeSyncStream` will be created and walk the whole subtree under this potentially large dir.
3. However not every step triggers an RPC to the UFS, because we do check again in `InodeSyncStream.processSyncPath()`.

So now the "extra work" is mostly the extra `InodeSyncStream` walk, which BFS walks the whole dir. During the walk, only the paths which are marked invalidated will really trigger UFS calls. Are there extra `UFS.listChildren` calls?

If the above is true, I don't really know how to estimate the extra cost. I mean the BFS walk is indeed non-trivial, but I don't know how that compares to UFS calls.

Reply:
If you are doing a non-recursive listing, you will only list and traverse your directory if an invalidation has been received on your immediate child.

If you are doing a recursive listing you will traverse the tree if any child has received an invalidation, but you will not traverse any child that has not received an invalidation.

Any folder/file on that path that has received an invalidation will trigger a UFS sync and traversal, but others will not.

For example if I have
```
                    dir1
          /           |       \
     /                |            \  
dir2            dir3             dir4
/    |         /     \             /    \
f1  f2        f3    f4           f5    f6
```

Now say they are all in sync and the following happens
- An invalidation is received on `/dir1/dir2/f0` (a new file has been created in dir2)
- A client performs `ls -R /dir1`
- The following will happen:
  - dir1 is listed on the UFS and its children are walked, a sync operation will be launched on dir2 since it is out of sync, but dir3, and dir4 will be skipped
  - dir2 is listed on the UFS and the children are walked, f1 and f2 will be skipped, and f0 will be created

This is the same basic idea as what happens now with the time based sync. If any individual child is out of sync, then each child has to be traversed that has a time that is out of sync.

Actually, I might be able to add an optimization to avoid the UFS listing of dir1, but I think all the other steps are necessary.

Reply:
Yea my point was also that the UFS listing on dir1 can be avoided theoretically. Feel free to do that in a separate PR :) Thanks for the graph that is super helpful.